src/System/Random/Stateful.hs

```diff
@@ -786,80 +787,97 @@ applyTGen f (TGenM tvar) = do
+-- * Otherwise:
+--
+--     1.  Generate an unsigned integral of matching width @__w__@ uniformly.
 --
--- If \(a = b\), return \(a\). Otherwise:
+--     2.  Check whether @__h - l__@ overflows to infinity and if does then convert @__w__@ to a floating
+--         point number in @__[0.0, 1.0]__@ range through division of @__w__@ by the highest possible value:
 --
--- 1.  Generate \(x\) uniformly such that \(0 \leq x \leq 1\).
+--         @
+--         x = `fromIntegral` w / `fromIntegral` `maxBound`
+--         @
 --
---     The method by which \(x\) is sampled does not cover all representable
---     floating point numbers in the unit interval. The method never generates
---     denormal floating point numbers, for example.
+--         Then we scale and clamp it before returning it:
 --
--- 2.  Return \(x \cdot a + (1 - x) \cdot b\).
+--         @
+--         `max` (`min` (x * l + (1 - x) * h) (`max` l h)) (`min` l h)
+--         @
 --
---     Due to rounding errors, floating point operations are neither
---     associative nor distributive the way the corresponding operations on
---     real numbers are. Additionally, floating point numbers admit special
---     values @NaN@ as well as negative and positive infinity.
+--         Clamping is necessary, because otherwise it would be possible to run into a
+--         degenerate case when a scaled value is outside the specified range due to
+--         rounding errors.
 --
--- For pathological values, step 2 can yield surprising results.
+--     3.  Whenever @__h - l__@ does not overflow, we use this common formula ofor scaling: @__ l + (h - l) * x__@.
+--         However, instead of using @__[0.0, 1.0]__@ range we use the top most bit of @__w__@ to decide whether
+--         we will treat the generated floating point value as @__[0.0, 0.5]__@ range or @__[0.5, 1.0]__@ range and use
+--         the left over bits to produce a floating point value in the half unit range:
 --
--- *   The result may be greater than @max a b@.
+--         @
+--         x = `fromIntegral` (`clearBit` w 31) / `fromIntegral` `maxBound`
+--         @
 --
---     >>> :{
---     let (a, b, x) = (-2.13238e-29, -2.1323799e-29, 0.27736077)
---         result = x * a + (1 - x) * b :: Float
---     in (result, result > max a b)
---     :}
---     (-2.1323797e-29,True)
+--         Further scaling depends on the top most bit:
 --
--- *   The result may be smaller than @min a b@.
+--         @
+--         if `testBit` w 31
+--            then l + (h - l) * x
+--            else h + (l - h) * x
+--         @
 --
---     >>> :{
---     let (a, b, x) = (-1.9087862, -1.908786, 0.4228573)
---         result = x * a + (1 - x) * b :: Float
---     in (result, result < min a b)
---     :}
---     (-1.9087863,True)
+--         Because of this clever technique the result does not need clamping, since
+--         scaled values are guaranteed to stay within the specified range.
```

Comment:
Maybe also note that this solution, in contrast to `x * l + (1 - x) * h`, strives for losing less randomness to rounding by avoiding calculation of `1 - x` and similar.

AFAICT, the current solution works equally well if `l < h` as well as `l > h` (always returning numbers between `l` and `h`). Not sure if this is worth mentioning.

Reply:
> Maybe also note that this solution,

Done

> Not sure if this is worth mentioning.

I don't think it is necessary. It was only mentioned before, because those instances for floating point values were violating the laws imposed on `uniformRM`:

> Generates a value uniformly distributed over the provided range, which is interpreted as inclusive in the lower and upper bound.